test/common/tracing/zipkin/zipkin_tracer_impl_test.cc

```diff
@@ -243,33 +243,114 @@ TEST_F(ZipkinDriverTest, FlushSpansTimer) {
+  EXPECT_TRUE(zipkin_span->span().sampled());
+}
+
+TEST_F(ZipkinDriverTest, NoB3ContextSampledFalse) {
+  setupValidDriver();
+
+  EXPECT_EQ(nullptr, request_headers_.XB3SpanId());
+  EXPECT_EQ(nullptr, request_headers_.XB3TraceId());
+  EXPECT_EQ(nullptr, request_headers_.XB3Sampled());
+
+  Tracing::SpanPtr span = driver_->startSpan(config_, request_headers_, operation_name_,
+                                             start_time_, {Tracing::Reason::Sampling, false});
+
+  ZipkinSpanPtr zipkin_span(dynamic_cast<ZipkinSpan*>(span.release()));
+  EXPECT_FALSE(zipkin_span->span().sampled());
+}
+
+TEST_F(ZipkinDriverTest, PropagateB3NoSampleDecisionSampleTrue) {
+  setupValidDriver();
+
+  request_headers_.insertXB3TraceId().value(Hex::uint64ToHex(Util::generateRandom64()));
+  request_headers_.insertXB3SpanId().value(Hex::uint64ToHex(Util::generateRandom64()));
+
+  Tracing::SpanPtr span = driver_->startSpan(config_, request_headers_, operation_name_,
+                                             start_time_, {Tracing::Reason::Sampling, true});
+
+  ZipkinSpanPtr zipkin_span(dynamic_cast<ZipkinSpan*>(span.release()));
+  EXPECT_TRUE(zipkin_span->span().sampled());
+}
+
+TEST_F(ZipkinDriverTest, PropagateB3NoSampleDecisionSampleFalse) {
+  setupValidDriver();
+
+  request_headers_.insertXB3TraceId().value(Hex::uint64ToHex(Util::generateRandom64()));
+  request_headers_.insertXB3SpanId().value(Hex::uint64ToHex(Util::generateRandom64()));
+
+  Tracing::SpanPtr span = driver_->startSpan(config_, request_headers_, operation_name_,
+                                             start_time_, {Tracing::Reason::Sampling, false});
+
+  ZipkinSpanPtr zipkin_span(dynamic_cast<ZipkinSpan*>(span.release()));
+  EXPECT_FALSE(zipkin_span->span().sampled());
+}
+
+TEST_F(ZipkinDriverTest, PropagateB3NotSampled) {
+  setupValidDriver();
+
+  EXPECT_EQ(nullptr, request_headers_.XB3SpanId());
+  EXPECT_EQ(nullptr, request_headers_.XB3TraceId());
+
+  // Only context header set is B3 sampled to indicate trace should not be sampled
+  request_headers_.insertXB3Sampled().value(ZipkinCoreConstants::get().NOT_SAMPLED);
+  Tracing::SpanPtr span = driver_->startSpan(config_, request_headers_, operation_name_,
+                                             start_time_, {Tracing::Reason::Sampling, true});
+
+  request_headers_.removeXB3Sampled();
+
+  span->injectContext(request_headers_);
+
+  // Check B3 sampled flag is set to not sample
+  EXPECT_TRUE(ZipkinCoreConstants::get().NOT_SAMPLED.compare(
+                  request_headers_.XB3Sampled()->value().getString()) == 0);
+}
+
+TEST_F(ZipkinDriverTest, PropagateB3NotSampledWithFalse) {
+  setupValidDriver();
+
+  EXPECT_EQ(nullptr, request_headers_.XB3SpanId());
+  EXPECT_EQ(nullptr, request_headers_.XB3TraceId());
+
+  // Only context header set is B3 sampled to indicate trace should not be sampled (using legacy
+  // 'false' value)
+  const std::string notSampled = "false";
```

Comment:
invert this condition. more important to test 'true' than 'false'

Reply:
added another test for 'true' case